### DISCUSSION TITLE
Other E2 KPM styles and E2 RC

### DISCUSSION THREAD

**Discussion by HaoxinSEU** [srsran/srsRAN_Project#459] [General Help] [answered]
Hi community,

The new update on E2 is really amazing! Following the tutorial, we can successfully run the gNB with FlexRIC, and the xApp can extract some E2-level metrics correctly.

Now we would like to move to UE-level metrics, which requires other KPM styles like style 4. However, we cannot run it successfully. Does anyone extract the per-UE metrics correctly by using other KPM styles?

BTW, does the branch `br-flexric` of FlexRIC support RC SM with srsRAN? We can't run the provided RC xApps from FlexRIC neither.

Any hints would be much appreciated!

**Comment by HaoxinSEU** [comment]
After digging into the RC, I found some problems. I'm using `xapp_oran_ctrl` in FlexRIC, to control the max and min PRBs (11 and 12 in style 2).

The first thing I did is to change the RAN parameter ID in `xapp_oran_ctrl.c`, because 4 is not in the spec.

But I still cannot run it correctly, so I try to trace the problem.

I find that the gNB successfully entered the function `e2sm_rc_control_action_2_6_du_executor::convert_to_du_config_request` in `/lib/e2/e2sm/e2sm_rc/e2sm_rc_control_action_du_executor.cpp`. But then `ctrl_msg.ran_p_list` **only contains RAN parameter 1**. So that the generated `ctrl_cfg` is empty, leading the function `e2sm_rc_control_action_2_6_du_executor::execute_ric_control_action` returns failure.

I'm not sure if this is a problem when gNB decodes the control message, or when xApp encodes the message? I don't change the generation function of CONTROL message in the xApp.

Does anyone have an idea how to proceed and debug this? Thanks

**Reply by HaoxinSEU** [reply]
The captured E2AP pcap is here. It's clear that E2 setup request and response are correct. In the CONTROL request, though wireshark cannot decode the content. But from the raw bits, I can figure out some values set in the xApp. For example, I can find the max_prb in the message. So I suppose the encoding in xApp may be correct?

[e2ap.zip](https://github.com/srsran/srsRAN_Project/files/14222025/e2ap.zip)

**Reply by peihaoY** [reply]
Hi HaoxinSEU,  can you run RAN Control (RC) xApp successfully now? I have the same issue. In the tutorial, for E2SM_RC service model, Control Service Style 2 is supported. But I cannot run `xapp_kpm_rc` and `xapp_oran_ctrl` .

**Reply by HaoxinSEU** [reply]
Hi @peihaoY,

Unfortunately I still cannot run the provided RC xApps in FlexRIC. We spent some time debugging the problem, and it's inside the ASN.1 decoding in gNB.

You can check this [thread ](https://github.com/srsran/srsRAN_Project/issues/468#issue-2129720728) for our investigation.

Let me know if I can provide more help.

**Reply by pgawlowicz** [reply, marked as answer]
Hi @HaoxinSEU and @peihaoY,

We have just released a repository that allows running ORAN-SC-RIC as a multi-container Docker application with a single docker command. You can access the repository at: https://github.com/srsran/oran-sc-ric

Moreover, the repository includes two Python-based xApps examples designed to facilitate the requesting and monitoring of E2 metrics from the gNB. These examples cover all reporting styles (1-5) of the e2sm-kpm module, which has been updated to version 3.

To get started quickly, refer to the `Quick Start` section in the readme file.
Please note that until the next code release, you'll need to utilize the `test` branch from the srsRAN_Project repository.

We are currently working on an example xApp using e2sm_rc module.

**Reply by HaoxinSEU** [reply]
Hi @pgawlowicz,

Thanks for the information! This helps a lot!

Thanks again,
Haoxin
